crates/wysiwyg/src/dom/nodes/mention_node.rs

```diff
@@ -22,47 +23,73 @@ use crate::dom::to_tree::ToTree;
 use crate::dom::unicode_string::{UnicodeStrExt, UnicodeStringExt};
 use crate::dom::UnicodeString;
 
+#[derive(Debug)]
+pub struct UriParseError;
+
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct MentionNode<S>
 where
     S: UnicodeString,
 {
-    kind: MentionNodeKind<S>,
+    // `display_text` refers to that passed by the client which may, in some cases, be different
+    // from the ruma derived `Mention.display_text`
+    display_text: S,
+    kind: MentionNodeKind,
     attributes: Vec<(S, S)>,
     handle: DomHandle,
 }
 
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub enum MentionNodeKind<S>
-where
-    S: UnicodeString,
-{
-    MatrixUrl { display_text: S, url: S },
+pub enum MentionNodeKind {
+    Room { mention: Mention },
+    User { mention: Mention },
```

Comment:
Could he type of mention Room / User be derived from the `mention` field? Otherwise we create the possibility of invalid states

E.g. `MentionNodeKind::Room(Mention { kind: MentionKind::User })`

I'd suggest keeping 

```
pub enum MentionNodeKind {
    MatrixUrl { mention: Mention },
    AtRoom,
}
```

Or perhaps better, integrating `AtRoom` into `matrix_mentions` crate.

Reply:
➕ on that last bit. 

When possible, things that apply the matrix mentions spec should be in that separate crate.

Reply:
Ok I have reverted it back to the MatrixURI style that it was before. I don't think that `AtRoom` should live in the mentions crate though - I'd argue it's more a node type as it contains none of the information that  MatrixURI mention would.